rate/rate.py

```diff
@@ -0,0 +1,228 @@
+        if self.user_vote != -1:
+            response = "Thank you for voting!"
+        else:
+            response = ""
+        rendered = html.format(self=self,
+                               scale=scale,
+                               freeform_prompt=prompt['freeform'],
+                               likert_prompt=prompt['likert'],
+                               response=response)
+
+        # We initialize self.p_r if not initialized -- this sets whether
+        # or not we show it. From there, if it is less than odds of showing,
+        # we set the fragment to the rendered XBlock. Otherwise, we return
+        # empty HTML. There ought to be a way to return None, but XBlocks
+        # doesn't support that.
+        if self.p_r == -1:
+            self.p_r = random.uniform(0, 100)
+        if self.p_r < self.p:
+            frag = Fragment(rendered)
+        else:
+            frag = Fragment(u"")
+
+        # Finally, we do the standard JS+CSS boilerplate. Honestly, XBlocks
+        # ought to have a sane default here.
+        frag.add_css(self.resource_string("static/css/rate.css"))
+        frag.add_javascript(self.resource_string("static/js/src/rate.js"))
+        frag.initialize_js('RateXBlock')
+        return frag
+
+    def studio_view(self, context):
+        """
+        Create a fragment used to display the edit view in the Studio.
+        """
+        html_str = pkg_resources.resource_string(__name__, "static/html/studio_view.html")
+        prompt = self.get_prompt(0)
+        frag = Fragment(unicode(html_str).format(**prompt))
+        js_str = pkg_resources.resource_string(__name__, "static/js/src/studio.js")
+        frag.add_javascript(unicode(js_str))
+        frag.initialize_js('RateBlock')
+        return frag
+
+    @XBlock.json_handler
+    def studio_submit(self, data, suffix=''):
+        """
+        Called when submitting the form in Studio.
+        """
+        self.prompts[0]['freeform'] = data.get('freeform')
+        self.prompts[0]['likert'] = data.get('likert')
+        return {'result': 'success'}
+
+    def vote(self, data):
+        """
+        Handle voting
+        """
+        # prompt_choice is initialized by student view.
+        # Ideally, we'd break this out into a function.
+        prompt = self.get_prompt(self.prompt_choice)
+
+        # Make sure we're initialized
+        if not self.vote_aggregate:
+            self.vote_aggregate = [0] * len(prompt['mouseovers'])
+
+        # Remove old vote if we voted before
+        if self.user_vote != -1:
+            self.vote_aggregate[self.user_vote] -= 1
+
+        self.user_vote = data['vote']
+        self.vote_aggregate[self.user_vote] += 1
+
+    @XBlock.json_handler
+    def feedback(self, data, suffix=''):
+        if 'freeform' in data:
+            tracker.emit('edx.ratexblock.freeform_feedback',
+                         {'old_freeform': self.user_freeform,
```

Comment:
Should add truncation. This is how we have been doing it--

The event includes both the old and the new values for the setting. 
If the string representation of a value is too long, it will be truncated and "..." will be added at the end
The event includes a "truncated" field which is an array of the fields that have been truncated, e.g. ["old"], ["old, "new"]

See https://openedx.atlassian.net/wiki/display/AN/User+Account+and+Preference+Event+Design.

It is my understanding that without truncation, a very long string could blow up the event log (impacting more than just the event data for this xblock).